### PR TITLE
Simplify (even more) list of characters to generate passwords

### LINKED ISSUE #102: Auto-generated passwords for postgres cannot contain % as special character
Not sure how serious this is but wanted to flag. 

when running kobo-install run.py -s passwords are autogenerated for postgres, mongo, and redis.

I ran into an issue where my automated backups were saving into an S3 bucket correctly and on schedule, but as 0 bytes. Digging through logs found that when then URI is created for postgres, if there are certain special characters in any of the username/password strings, postgres will complain and not succeed in a dump. 

This happened to me with an autogenerated pw that included a %  char, but I imagine this would happen with @ as well. Obviously one could just change the password manually which is what I ended up doing, but was a short-lived mystery that the backups werent completeing correctly, and only found out because I check those things.

Perhaps a note on which characters not to use for the uname and pwd during the setup process?  or a check during setup would be helpful?

### PR DESCRIPTION
**Pull request by noliveleger**
## Description
Some characters (e.g. `^`, `$`, `#` ) are problematic when used in PostgreSQL, MongoDB or Redis passwords.
When running `--setup` (or `--update`), password validation blocks user's input if it contains such characters. 
Moreover, the passwords are validated on start in case of an old config would contain unsupported characters (e.g the 
 user has updated their kobo-install environment with `git pull` instead of `./run --update`).
 A warning is displayed in the console if it's the case.

## Additional info
DB password encoding has been brought back because it was removed in #182 by mistake. 

## Related issues
Closes #102

### CODE REVIEW

**Comment by bufke** [comment]
What's the need for POSTGRES_PASSWORD_URL_ENCODED? I see django-environ [mentions it](https://django-environ.readthedocs.io/en/latest/tips.html#using-unsafe-characters-in-urls). I think it would be nicer to have less environment variables and more consistency. The Redis password doesn't do this. "There should be one-- and preferably only one --obvious way to do it."

I tested your allowed special variables and they all seem to work without this quoting them.

```
root@7bda3309b04a:/code# env | grep DATABASE
DATABASE_URL=postgres://postgres:.!+_~`-pass@postgres:5432/postgres
root@7bda3309b04a:/code# ./manage.py migrate
 (works)
```

**Comment by noliveleger** [comment]
> What's the need for POSTGRES_PASSWORD_URL_ENCODED? I see django-environ [mentions it](https://django-environ.readthedocs.io/en/latest/tips.html#using-unsafe-characters-in-urls). I think it would be nicer to have less variables and more consistency. The Redis password doesn't do this. "There should be one-- and preferably only one --obvious way to do it."
> 
> I tested your allowed special variables and they all seem to work without quoting them.
> 
> ```
> root@7bda3309b04a:/code# env | grep DATABASE
> DATABASE_URL=postgres://postgres:.!+_~`-pass@postgres:5432/postgres
> root@7bda3309b04a:/code# ./manage.py migrate
>  (works)
> ```


Does not for me. 
```
In [1]: import os
In [2]: os.environ['TEST_URL'] = 'postgres://postgres:.!+_~`-pass@postgres:5432/postgres'
In [3]: import environ
In [4]: env = environ.Env()
In [5]: env.db_url('TEST_URL')
Out[5]:
{'NAME': 'postgres',
 'USER': 'postgres',
 'PASSWORD': '.! _~`-pass',  # Plus sign has been replace by space
 'HOST': 'postgres',
 'PORT': 5432,
 'ENGINE': 'django.db.backends.postgresql'}
In [6]:
 ```

I used `.db_url()` but it does the same with `.db()`

> The Redis password doesn't do this. There should be one-- and preferably only one --obvious way to do it.

I agree. but IMHO the guilty guy is `django-environ` at the first place which does not treat passwords in `db_url` and `cache_url` (or `.db()` and `.cache()`) the same way.  
FYI: Mongo is also using a different way because it does not use a connection string. It's built inside KPI and KoBoCAT settings.

**Comment by bufke** [comment]
Ah I see, in my example I was merely consistently turning + into space. Which is indeed quite horrible but strangely works.

I'm still tempted to say if we are talking about just one more special character - remove it from the list of valid chars and simplify the install scripts. It's typical in the kube community (see defaults in postgresql helm chart) to generate passwords with only alphanumeric values of length 10 and it's plenty of entropy. We use 20 chars, so the extra bits of entropy added by additional special chars is irrelevant. Take my opinion it or leave it. If we just want special chars and have a history of supporting most of them, sure. The main argument IMO for special chars is to try to get human typed passwords to be better, but these are not human typed passwords.

I tested this branch out and was able to trigger the Invalid password on an existing install. The message was clear to me and I was able to update my instance without a problem. This looks good to merge to me.

**Comment by noliveleger** [comment]
> We use 20 chars, so the extra bits of entropy added by additional special chars is irrelevant. Take my opinion it or leave it

@bufke, I do take your wisedom into consideration. @jnm shares the same by the way. I just wanted to keep some special characters for retro-compatibility with servers we are managing but some characters would have to be stripped anyway. 
There is a consensus, so let's keep this path then.

File: helpers/config.py [hunk 1]
```diff
@@ -581,6 +581,59 @@ def validate_aws_credentials(self):
...
+            },
+        }
+        errors = []
+        for label, config_ in passwords.items():
+            if not re.match(config_['pattern'], config_['password']):
+                errors.append(label)
+                CLI.colored_print(
+                    f'{label} password contains unsupported characters.',
+                    CLI.COLOR_WARNING
+                )
```
Comment (joshuaberetta) [inline helpers/config.py:618]: Maybe rather something like this? But it's not really important...

```python3

def get_pattern(chars=8, allow_empty=False):
  pattern = f'[{self.ALLOWED_PASSWORD_CHARACTERS}]{{{chars},}}'
  if allow_empty:
    pattern += '|'
  return rf'^{pattern}$'

...

passwords = {
  ...
  'redis': {
    'label': 'Redis',
    'pattern': get_pattern(allow_empty=True)
  }
}
errors = []
for name, config_ in passwords.items():
  if not re.match(config_['pattern'], self.__dict[f'{name}_password']):
    errors.append(name)
    CLI.colored_print(
      f'{config_['label']} password contains unsupported characters.',
      CLI.COLOR_WARNING
    )
```

**Review by joshuaberetta — CHANGES_REQUESTED** [review]

File: helpers/config.py (hunk 1 above)
Comment (noliveleger) [reply helpers/config.py:618]: Great idea ;-)

**Pull request merged**